docs/architetture-software/README.md

```diff
@@ -0,0 +1,212 @@
+- **Scalabilità**: 1. Il monolite, anche in forma modulare, è l'architettura software più difficile da scalare, in quanto non è possibile scalare i singoli componenti, ma è necessario scalare l'intero sistema. In questi casi, infatti, al monolite viene affiancato il concetto di _scalabilità verticale_, ovvero l'incremento delle risorse hardware a disposizione del sistema (CPU, RAM, storage, ecc.) quando queste non sono più sufficienti per gestire il carico di lavoro. La valutazione è dovuta alla necessità, nella maggior parte delle implementazioni di questo tipo, di dover riavviare il sistema per scalare le risorse hardware.
+- **Performance**: 1. La performance del monolite modulare presenta le stesse problematiche del monolite _classico_, con l'aggiunta che per far funzionare questa architettura è necessario un ulteriore livello di astrazione, ovvero il modulo, che può portare ad un degrado delle performance.
+
+## Microkernel
+
+Chiamata anche _architettura a plugin_, si compone essenzialmente di un nucleo centrale che si occupa di gestire le comunicazioni tra i vari componenti, che sono sviluppati e distribuiti come plugin.
+Questi plugin possono essere sviluppati e distribuiti in maniera indipendente, ma funzionando attorno ad un nucleo centrale, faranno sempre parte dello stesso _blocco_, inteso che l'applicativo online è unico, come nel caso del monolite _classico_.
+
+Il concetto principale attorno a questa architettura è l'indipendenza totale tra i plugin. Per definizione, nessun plugin dovrebbe avere come dipendenza un altro plugin.
+
+In questa architettura il database, il back-end e il front-end sono tutti parte dello stesso processo/applicativo.
+
+- **Start-up**: 4. Essendo una variante del monolite, lo start-up è molto semplice, ma la necessità di organizzare il codice in un nucleo centrale e implementare un sistema di plugin può richiedere più tempo.
+- **Costi**: 3. Pur essendo un'architettura _a costo zero_ in termini di start-up, il costo cresce esponenzialmente per via della necessità di _scalare orizzontalmente_.
+- **Sviluppo e manutenzione**: 4. Trattasi di un'architettura semplice da mantenere e sul quale aggiungere funzionalità, ma la necessità di non far comunicare i plugin tra di loro può aggiungere una lieve complessità.
+- **Tracciabilità**: 3. Testare e tracciare eventuali bug in questa architettura è mediamente complesso. L'isolamento dei plugin aiuta con il testing, ma la correlazione col nucleo centrale a volte può rallentare il processo di tracciamento.
+- **Scalabilità**: 2. Le difficoltà di scalabilità sono anche in questo caso dovute alla _singola unità di rilascio_ o _single deployment unit_, che non ne permette la _scalabilità orizzontale_.
+- **Performance**: 4. Processi di comunicazione semplici e isolabilità rendono questa architettura molto performante.
+
+## Microservizi
```

Comment:
Spiegherei che i ms non sono una soluzione tecnica ma organizzativa. Dal punto di vista tecnico, portano solo problemi, facendo diventare la nostra applicazione,  un sistema dostribuito. I vantaggi principali che si vogliono ottenere sono independent deployability e il decoupling. Separare in team per avere una separazione di velocity e poter deploiare in maniera indipendente. Il disaccoppiamento invece permette di ridurre la complessiva che viene a crearsi nei monoliti fatti male. Ovvero entita che rappresentano le esigenze di piu' team e che quindi diventano presto un inferma da mantenere e far evolvere

Reply:
@frnmjn anche questa è più un'opinione che un fatto. Anche il monolite possiamo dirci che è una tecnica organizzativa.. Nessuna architettura ha delle _capacità_ tecniche che le altre architetture non hanno, non trovi?

Ho aggiunto il seguente paragrafo, dimmi se è un buon compromesso per te:

Per definizione i microservizi non dovrebbero comporsi di più di una manciata di endpoint e relative funzioni, in quanto altrimenti si parla più di _service-based architecture_.

Da notare che il _decoupling_ creato da questa architettura può portare ad un notevole aumento della complessità.

Reply:
sulle dimensioni di un microservizio, non ce una metrica. Ci sono varie definizioni qui https://www.ben-morris.com/how-big-is-a-microservice/#:~:text=For%20example%2C%20Sam%20Newman%20described,than%20a%20handful%20of%20people%E2%80%9D.

Reply:
io facendo DDD, mappo un microservizio con un bounded context con 1 o piu' aggragati. Una definizione che mi piace di microservizio e': grande finche ci sta nella testa di un singolo sviluppatore

Reply:
Potrebbe magari aver senso aggiungere un paragrafo di "opinioni comuni", che ne pensi? Proponi una PR così la butti giù nel modo che preferisci, che ne dici?